### PR TITLE
Create Toggle

### PR DESCRIPTION
**Pull request by Twinters007**
Like debounce, just another useful feature for buttons. I added Toggle.on and Toggle.off to remove boilerplate when only wanting to perform an action during one state

```
if toggle.on:
    toggleOnFoo()
```
vs

```
on = False
if toggle and not on:
    toggleOnFoo()
    on = True
if toggle and on:
    on = False
```

I also allowed the option of adding a debounce to the toggle. I had to create a new debounce because we want the value to remain true even if the joystick bounces, instead of returning true once, then false until the time expires

### CODE REVIEW

File: robotpy_ext/control/toggle.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        """
+        :param joystick: wpilib.Joystick that contains the button to toggle
+        :param button: Value of button that will act as toggle. Same value used in getRawButton()
+        """
+
+        if debouncePeriod is not None:
+            self.joystick = Toggle._SteadyDebounce(joystick, button, debouncePeriod)
+        else:
+            self.joystick = joystick
+            self.joystick.get = partial(self.joystick.getRawButton, button)
```
Comment (Twinters007) [inline robotpy_ext/control/toggle.py]: I decided to use a partial here because then I don't need two copies of the code depending on whether or not there's a debounce. And I learned what a partial is

File: robotpy_ext/control/toggle.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    @property
+    def on(self):
+        return self.state == 1
+
+    @property
+    def off(self):
+        return self.state == 0
+
+    __bool__ = get
```
Comment (Twinters007) [inline robotpy_ext/control/toggle.py]: Just so they can use the toggle like a boolean in an if statement

File: robotpy_ext/control/toggle.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            Similar to ButtonDebouncer, but the output stays steady for
+            the given periodic_filter. E.g, if you set the period to 2
+            and press the button, the value will return true for 2 seconds.
+
+            Steady debounce will return true for the given period, allowing it to be
+            used with Toggle
+        '''
+
+        def __init__(self, joystick, button, period=0.5):
+            from wpilib.timer import Timer
```
Comment (virtuald) [inline robotpy_ext/control/toggle.py]: Why is this here instead of the top of the file?

File: robotpy_ext/control/toggle.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            '''Set number of seconds to hold return value'''
+            self.debounce_period = float(period)
+
+        def get(self):
+            '''Returns the value of the joystick button. Once the button is pressed,
+            the return value will be True until the time expires
+            '''
+
+            now = self.timer.getFPGATimestamp()
+            if now - self.latest < self.debounce_period:
```
Comment (virtuald) [inline robotpy_ext/control/toggle.py]: This seems wrong. What am I missing here?

File: robotpy_ext/control/toggle.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        if foo:
+            toggleFunction()
+
+        if foo.on:
+            onToggle()
+
+        if foo.off:
+            offToggle()
+    """
+    class _SteadyDebounce:
```
Comment (virtuald) [inline robotpy_ext/control/toggle.py]: Why is this an inner class?

File: robotpy_ext/control/toggle.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        """
+         :return: State of toggle
+         :rtype: bool
+         """
+        current_state = self.joystick.get()
+
+        if current_state and not self.released:
+            self.released = True
+            self.toggle = not self.toggle
+            self.state = 1 - self.state
```
Comment (virtuald) [inline robotpy_ext/control/toggle.py]: Clever. Needs a comment though.

File: robotpy_ext/control/toggle.py (hunk 4 above)
Comment (Twinters007) [inline robotpy_ext/control/toggle.py]: Because I want it to return true while the time is less than the period. 

Let's say `debounce_period = 1`. If I press the button at `t = 0`, then at `t = .5` it will still return true.

File: robotpy_ext/control/toggle.py (hunk 3 above)
Comment (Twinters007) [inline robotpy_ext/control/toggle.py]: Not sure what the general consensus is with importing when you need to vs just doing it anyway. Should it be at the top?

File: robotpy_ext/control/toggle.py (hunk 5 above)
Comment (Twinters007) [inline robotpy_ext/control/toggle.py]: I didn't think it'd be useful except for this toggle feature.

File: robotpy_ext/control/toggle.py (hunk 3 above)
Comment (virtuald) [inline robotpy_ext/control/toggle.py]: Yes.

**Comment by virtuald** [comment]
I would be happier if this had some associated tests with it.

**Comment by Twinters007** [comment]
How do I simulate joysticks in robotpy_ext?

**Comment by virtuald** [comment]
Could use a fake class.

File: robotpy_ext/control/toggle.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        """
+         :return: State of toggle
+         :rtype: bool
+         """
+        current_state = self.joystick.get()
+
+        if current_state and not self.released:
+            self.released = True
+            self.toggle = not self.toggle
+            self.state = 1 - self.state  # Toggles between 1 and 0.
```
Comment (auscompgeek) [inline robotpy_ext/control/toggle.py]: Why isn't this a boolean instead?

File: robotpy_ext/control/toggle.py (hunk 5 above)
Comment (auscompgeek) [inline robotpy_ext/control/toggle.py]: Hmm. Is debouncing actually useful for a button toggle?

File: robotpy_ext/control/toggle.py (hunk 7 above)
Comment (Twinters007) [inline robotpy_ext/control/toggle.py]: Codewise it makes more sense for 1 or 0 because it's not really True or False, but a boolean would do the same thing

Comment (auscompgeek) [inline robotpy_ext/control/toggle.py]: I don't think it does. A toggle is either on or off, right? True or false.

Comment (Twinters007) [inline robotpy_ext/control/toggle.py]: yeah idk. I can change

File: robotpy_ext/control/toggle.py (hunk 5 above)
Comment (Twinters007) [inline robotpy_ext/control/toggle.py]: It's as useful as a normal debounce I guess? If you think you need it, you can put it in, if you don't, leave it out.

File: tests/test_toggle.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    joystick.release()
+    assert toggleButton.on
+    joystick.press()
+    assert toggleButton.off
+    joystick.release()
+    assert toggleButton.off
+
+
+def test_toggle_debounce():
+    delay = PreciseDelay(2.1)
```
Comment (Twinters007) [inline tests/test_toggle.py]: Setting the precise delay to the exact timeout sometimes didn't work.

**Comment by virtuald** [comment]
If you fix the conflicts and squash the commits I'll merge this.
